lib/media/streaming_engine.js

```diff
@@ -317,6 +321,88 @@ shaka.media.StreamingEngine = class {
+      shaka.log.debug('variant missing, no switch');
+      // One of the variants is null or missing
+    } else {
+      const ContentType = shaka.util.ManifestParserUtils.ContentType;
+      const audioCodecsMatch = this.doVariantCodecsMatchByContentType(
+          previousVariant, currentVariant, ContentType.AUDIO);
+      const videoCodecsMatch = this.doVariantCodecsMatchByContentType(
+          previousVariant, currentVariant, ContentType.VIDEO);
+      if (!audioCodecsMatch) {
+        shaka.log.debug('Audio Codec Switch');
+        returnValue.push(ContentType.AUDIO);
+      }
+      if (!videoCodecsMatch) {
+        shaka.log.debug('Video Codec Switch');
+        returnValue.push(ContentType.VIDEO);
+      }
+    }
+    return returnValue;
+  }
+
+  /**
+   * Determine if Codec match by ContentType.
+   * @param {shaka.extern.Variant} previousVariant
+   * @param {shaka.extern.Variant} currentVariant
+   * @param {shaka.util.ManifestParserUtils.ContentType} contentType
+   * @return {boolean}
+   */
+  doVariantCodecsMatchByContentType(previousVariant,
+      currentVariant, contentType) {
+    const pCodecs = previousVariant[contentType] &&
+    previousVariant[contentType].codecs;
+    const cCodecs = currentVariant[contentType] &&
+    currentVariant[contentType].codecs;
+    return (pCodecs && cCodecs && pCodecs === cCodecs);
+  }
+
+  /**
+   * Enforce Codec Switch Strategy by Content Type.
+   * @param {shaka.config.CodecSwitchingStrategy} strategy
+   * @param {Array.<shaka.util.ManifestParserUtils.ContentType>} contentTypes
+   */
+  async enforceCodecStrategyByContentTypes(strategy, contentTypes) {
+    const ContentType = shaka.util.ManifestParserUtils.ContentType;
+    const CodecSwitchingStrategy = shaka.config.CodecSwitchingStrategy;
+    if (strategy === CodecSwitchingStrategy.SMOOTH) {
+      shaka.log.debug(`Attempting ${CodecSwitchingStrategy.SMOOTH} Strategy`);
+      if (shaka.media.MediaSourceEngine.isChangeTypeSupported()) {
+        // Supported Go Forth and change!
+        contentTypes.forEach(async (v, i) => {
+          const contentType = v;
+          const variant = this.currentVariant_;
+          const fullMimeType = shaka.util.MimeUtils.getFullType(
+              variant[contentType].mimeType, variant[contentType].codecs);
+          await this.playerInterface_.mediaSourceEngine.changeType(
+              contentType, fullMimeType);
+        });
+        return;
+      } else {
+        shaka.log.debug(`${CodecSwitchingStrategy.SMOOTH}` +
+        `Strategy not supported, fallback to ${CodecSwitchingStrategy.RELOAD}`);
+        // Not Supported, fallback to RELOAD
+      }
+    }
+
+    // RELOAD Strategy
+    shaka.log.debug(`Using ${CodecSwitchingStrategy.RELOAD} Strategy`);
+    this.unloadMediaStateByContentType(ContentType.AUDIO);
+    this.unloadMediaStateByContentType(ContentType.VIDEO);
+    await this.playerInterface_.mediaSourceEngine
+        .reset(contentTypes[0]);
```

Comment:
What about the rest of the contentTypes?

Reply:
On reset, the contentType really is just used to fired the fake event at the end. The audio and video are ultimately reset together regardless of what is input here. I can look to fire that with null if need be.